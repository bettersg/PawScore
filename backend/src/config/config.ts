import dotenv from "dotenv";
import path from "path";

const env = process.env.NODE_ENV || "development";

// eslint-disable-next-line
const config: { [index: string]: any } = new Map<string, any>();

dotenv.config({ path: path.resolve(__dirname + "/../../.env") });

config["development"] = {
	databaseConfig: {
		username: process.env.DB_USER,
		password: process.env.DB_PASS,
		database: process.env.DB_DATABASE,
		host: process.env.DB_HOST,
		dialect: "postgres",
		define: {
			underscored: true,
			underscoredAll: true
		}
	},
	cookieSecret: process.env.COOKIE_SECRET || "secrete",
	expressHost: process.env.EXPRESS_HOST || "127.0.0.1",
	expressPort: 5000,
	storageBucketName: process.env.STORAGE_BUCKET_NAME,
	googleSvcAcctKey: process.env.GOOGLE_SVC_ACCT_KEY,
	frontendUrl: process.env.FRONTEND_URL || "http://127.0.0.1:3000"
};

config["test"] = {
	databaseConfig: {
		username: process.env.DB_USER,
		password: process.env.DB_PASS,
		database: process.env.DB_DATABASE,
		host: process.env.DB_HOST,
		dialect: "postgres",
		define: {
			underscored: true,
			underscoredAll: true
		}
	},
	cookieSecret: process.env.COOKIE_SECRET,
	expressHost: process.env.EXPRESS_HOST || "127.0.0.1",
	expressPort: 5000,
	storageBucketName: process.env.STORAGE_BUCKET_NAME,
	googleSvcAcctKey: process.env.GOOGLE_SVC_ACCT_KEY,
	frontendUrl: process.env.FRONTEND_URL || "http://127.0.0.1:3000"
};

config["production"] = {
	databaseConfig: {
		username: process.env.DB_USER,
		password: process.env.DB_PASS,
		database: process.env.DB_DATABASE,
		host: process.env.DB_HOST,
		dialect: "postgres",
		define: {
			underscored: true,
			underscoredAll: true
		}
	},
	cookieSecret: process.env.COOKIE_SECRET,
	expressHost: process.env.EXPRESS_HOST || "0.0.0.0",
	expressPort: 5000,
	storageBucketName: process.env.STORAGE_BUCKET_NAME,
	googleSvcAcctKey: process.env.GOOGLE_SVC_ACCT_KEY,
	frontendUrl: process.env.FRONTEND_URL || "http://127.0.0.1:3000"
};

export default config[env];
