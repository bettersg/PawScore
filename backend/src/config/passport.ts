import bCrypt from "bcrypt";
import express from "express";
import passportLocal from "passport-local";
import passport from "passport";
import { User, UserCreationAttributes } from "../models/user";

const authStrategy = (passport: passport.PassportStatic): void => {
	const LocalStrategy = passportLocal.Strategy;
	// serialize
	passport.serializeUser(async (_user, done) => {
		const user = _user as unknown as User;
		done(null, user?.id);
	});

	// deserialize user
	passport.deserializeUser(async (id: string, done) => {
		try {
			const user = await User.findOne({ where: { id } });
			if (user === null) {
				throw new Error(`Error deserializing user with id ${id}`);
			}
			done(null, user.get());
		} catch (err) {
			done(err, null);
		}
	});

	passport.use(
		"local-signup",
		new LocalStrategy(
			{
				usernameField: "email",
				passwordField: "password",
				passReqToCallback: true // allows us to pass back the entire request to the callback
			},
			async (
				req: express.Request,
				username: string,
				password: string,
				done
			) => {
				try {
					const generateHash = (genPass: string) => {
						return bCrypt.hashSync(genPass, bCrypt.genSaltSync(8));
					};

					const existingUser = await User.findOne({
						where: {
							email: req.body.email
						}
					});
					if (existingUser) {
						return done(Error("Username taken"), null, {
							message: "That email address is already taken"
						});
					} else {
						const userPasswordHash = generateHash(password);
						const data: UserCreationAttributes = {
							username: req.body.username || req.body.email,
							email: req.body.email,
							password: userPasswordHash,
							roles: ["ADOPTER"],
							shelterId: null
						};
						const newUser = await User.create(data);
						if (!newUser) {
							return done(null, null);
						}
						if (newUser) {
							return done(null, newUser);
						}
					}
				} catch (err) {
					done(err, null);
				}
			}
		)
	);
	passport.use(
		"local-login",
		new LocalStrategy(
			{
				usernameField: "email",
				passwordField: "password",
				passReqToCallback: true
			},
			async (
				req: express.Request,
				username: string,
				password: string,
				done
			) => {
				try {
					const userFound = await User.findOne({
						where: {
							email: req.body.email
						}
					});
					if (userFound === null) {
						return done(null, null, {
							message: "Incorrect email address."
						});
					} else if (
						!bCrypt.compareSync(password, userFound.password)
					) {
						return done(null, null, {
							message: "Incorrect password."
						});
					}
					return done(null, userFound);
				} catch (err) {
					done(err, null);
				}
			}
		)
	);
};

export default authStrategy;
