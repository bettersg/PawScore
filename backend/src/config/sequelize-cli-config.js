// this file is used mainly for migration only
const path = require("path");
require('dotenv').config({path: path.resolve(__dirname + "/../../.env")}); // this is important!
console.log(path.resolve(__dirname + "/../../.env"));
console.log(process.env.DB_USER);

module.exports = {
  "development": {
      "username": process.env.DB_USER,
      "password": process.env.DB_PASS,
      "database": process.env.DB_DATABASE,
      "host": process.env.DB_HOST,
      "dialect": "postgres"
  },
  "test": {
      "username": process.env.DB_USER,
      "password": process.env.DB_PASS,
      "database": process.env.DB_DATABASE,
      "host": process.env.DB_HOST,
      "dialect": "postgres"
  },
  "production": {
      "username": process.env.DB_USER,
      "password": process.env.DB_PASS,
      "database": process.env.DB_DATABASE,
      "host": process.env.DB_HOST,
      "dialect": "postgres"
  }
};
