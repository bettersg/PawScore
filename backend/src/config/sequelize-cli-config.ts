// this file is used mainly for migration only
import path from "path"
import dotenv from "dotenv"
dotenv.config({path: path.resolve(__dirname + "/../../.env")});

import fullConfig from "./config"

module.exports = fullConfig.databaseConfig;
