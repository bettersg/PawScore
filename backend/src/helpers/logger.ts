import { AsyncLocalStorage, AsyncResource } from "async_hooks";
import crypto from "crypto";
import { Request, RequestHandler } from "express";

type RequestContext = {
	id: string;
	req: Request;
};

// set up a store to persist data through async operations, making them accessible from nested calls
export const requestContextLocalStorage =
	new AsyncLocalStorage<RequestContext>();

export class Logger {
	static log(...message: any[]) {
		const context = requestContextLocalStorage.getStore();
		console.log({
			context,
			message,
		});
	}
}

export const LoggerContextMiddleware: RequestHandler = (req, res, next) => {
	const context = {
		id: crypto.randomUUID(),
		req,
	};

	requestContextLocalStorage.run(context, () => {
		Logger.log("Request start", {
			body: req.body,
			query: req.query,
		});
		// successful responses strangely do not emit `finish`, use `close`
		res.on(
			"close",
			// binding required as context was lost
			AsyncResource.bind(() => {
				Logger.log("Request end", { status: res.statusCode });
			}),
		);
		next();
	});
};
