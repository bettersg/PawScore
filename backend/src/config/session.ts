import pg from "pg";
import expressSession from "express-session";
import connectPgSimple from "connect-pg-simple";
import config from "./config";
import express from "express";

const pgSession = connectPgSimple(expressSession);

const pgPool = new pg.Pool({
  host: config.databaseConfig.host,
  user: config.databaseConfig.username,
  database: config.databaseConfig.database,
  password: config.databaseConfig.password,
  max: 10,
  idleTimeoutMillis: 30000,
  connectionTimeoutMillis: 2000,
});

export default (app: express.Application): void => {
  app.use(expressSession({
    store: new pgSession({
      pool : pgPool,                // Connection pool
      tableName : 'user_sessions'   // Use another table-name than the default "session" one
      // Insert connect-pg-simple options here
    }),
    secret: config.cookieSecret,
    cookie: { maxAge: 30 * 24 * 60 * 60 * 1000 }, // 30 days
    // Insert express-session options here
    saveUninitialized: false,
    resave: false,
  }));
};
