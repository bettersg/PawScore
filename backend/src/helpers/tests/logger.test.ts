import { Request } from "express";
import { BadRequestError } from "routing-controllers";
import config from "../../config/config";
import { Logger, requestContextLocalStorage } from "../logger";

jest.mock("../../config/config", () => ({
	__esModule: true,
	default: {
		nodeEnv: null,
		logMaxStringLength: 100,
		logLevel: "debug",
	},
}));

const consoleSpy = jest.spyOn(global.console, "log");

describe("Logger", () => {
	beforeAll(() => {
		jest.useFakeTimers("modern");
		jest.setSystemTime(new Date("2022-01-31T16:00:00.000Z"));
	});

	afterAll(() => {
		jest.useRealTimers();
	});

	beforeEach(() => {
		config.nodeEnv = "";
		Logger.setLogLevel("debug");
		jest.clearAllMocks();
	});

	test("should default to printing single line JSON", () => {
		Logger.info("message", {
			a: "a",
			b: "b",
		});

		expect(consoleSpy.mock.calls[0][0]).toEqual(
			`{"req":{},"timestamp":"2022-01-31T16:00:00.000Z","severity":"INFO","message":"message","data":[{"a":"a","b":"b"}]}`,
		);
	});

	test("should pretty-print JSON in local environment", () => {
		config.nodeEnv = "development";
		Logger.info("message", {
			a: "a",
			b: "b",
		});

		expect(consoleSpy.mock.calls[0][0]).toEqual(
			`{\n  "req": {},\n  "timestamp": "2022-01-31T16:00:00.000Z",\n  "severity": "INFO",\n  "message": "message",\n  "data": [\n    {\n      "a": "a",\n      "b": "b"\n    }\n  ]\n}`,
		);
	});

	test("should log on equal or higher severity", () => {
		Logger.setLogLevel("error");

		Logger.debug("message");
		Logger.info("message");
		Logger.error("message");
		Logger.fatal("message");

		expect(consoleSpy.mock.calls).toHaveLength(2);
		expect(consoleSpy.mock.calls).toMatchSnapshot();
	});

	test("should extract req information when available", () => {
		const body = {
			a: "a",
			b: "b",
		};

		const context = {
			id: "reqId",
			req: {
				path: "reqPath",
				header: (name: string) => {
					return name === "X-Cloud-Trace-Context"
						? "105445aa7843bc8bf206b12000100000/1;o=1"
						: "reqHeader";
				},
			} as Request,
		};

		requestContextLocalStorage.run(context, () => {
			Logger.info("message", body);
		});

		expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot();
	});

	test("should redact blacklisted fields", () => {
		const body = {
			password: "abcdef",
			notPassword: "123456",
		};

		Logger.info("message", body);

		expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot();
	});

	test("should truncate long strings", () => {
		const body = {
			// 110 chars
			long: new Array(11)
				.fill(0)
				.map(() => "0123456789")
				.join(""),
			nested: {
				long: new Array(11)
					.fill(0)
					.map(() => "0123456789")
					.join(""),
			},
			// 100 chars
			short: new Array(10)
				.fill(0)
				.map(() => "0123456789")
				.join(""),
		};

		Logger.info("message", body);

		expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot();
	});

	test("should print error", () => {
		const error = new Error("error message");

		Logger.info("message", error);

		expect(consoleSpy.mock.calls[0][0]).toEqual(
			`{"req":{},"timestamp":"2022-01-31T16:00:00.000Z","severity":"INFO","message":"message","data":[{"errorName":"Error","errorMessage":"error message"}]}`,
		);
	});

	test("should print routing-controllers error", () => {
		const error = new BadRequestError("error message");

		Logger.info("message", error);

		expect(consoleSpy.mock.calls[0][0]).toEqual(
			`{"req":{},"timestamp":"2022-01-31T16:00:00.000Z","severity":"INFO","message":"message","data":[{"errorName":"BadRequestError","errorMessage":"error message"}]}`,
		);
	});
});
