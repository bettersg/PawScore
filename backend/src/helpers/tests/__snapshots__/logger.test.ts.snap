// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should extract req information when available 1`] = `"{\\"req\\":{\\"url\\":\\"reqPath\\",\\"id\\":\\"reqId\\"},\\"timestamp\\":\\"2022-01-31T16:00:00.000Z\\",\\"severity\\":\\"INFO\\",\\"logging.googleapis.com/trace\\":\\"projects/pawscore/traces/105445aa7843bc8bf206b12000100000\\",\\"message\\":\\"message\\",\\"data\\":[{\\"a\\":\\"a\\",\\"b\\":\\"b\\"}]}"`;

exports[`Logger should log on equal or higher severity 1`] = `
Array [
  Array [
    "{\\"req\\":{},\\"timestamp\\":\\"2022-01-31T16:00:00.000Z\\",\\"severity\\":\\"ERROR\\",\\"message\\":\\"message\\",\\"data\\":[]}",
  ],
  Array [
    "{\\"req\\":{},\\"timestamp\\":\\"2022-01-31T16:00:00.000Z\\",\\"severity\\":\\"CRITICAL\\",\\"message\\":\\"message\\",\\"data\\":[]}",
  ],
]
`;

exports[`Logger should redact blacklisted fields 1`] = `"{\\"req\\":{},\\"timestamp\\":\\"2022-01-31T16:00:00.000Z\\",\\"severity\\":\\"INFO\\",\\"message\\":\\"message\\",\\"data\\":[{\\"password\\":\\"[Redacted]\\",\\"notPassword\\":\\"123456\\"}]}"`;

exports[`Logger should truncate long strings 1`] = `"{\\"req\\":{},\\"timestamp\\":\\"2022-01-31T16:00:00.000Z\\",\\"severity\\":\\"INFO\\",\\"message\\":\\"message\\",\\"data\\":[{\\"long\\":\\"0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789[...]\\",\\"nested\\":{\\"long\\":\\"0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789[...]\\"},\\"short\\":\\"0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\\"}]}"`;
